lib/cretonne/src/licm.rs

```diff
@@ -15,78 +22,189 @@ pub fn do_licm(func: &mut Function,
+    opcode.can_trap() || opcode.other_side_effects()
+}
+
+// Helper function to move an instruction inside a function
+fn move_inst(layout: &mut Layout, to_move: Inst, destination: Inst) {
+    let mut cur = Cursor::new(layout);
+    cur.goto_inst(to_move);
+    cur.remove_inst();
+    cur.goto_inst(destination);
+    cur.insert_inst(to_move);
+}
+
+// Given a instruction and the inner-most loop it is part of, determine from the positions of its
+// arguments the closest definition of one of the instruction's argument. This closest position
+// can be:
+// - inside the `inst_loop`, meaning that the instruction is not loop-invariant;
+// - inside a parent loop which is not `inst_loop`.
+// - outside any loop;
+fn closest_def_position(func: &Function,
+                        inst: Inst,
+                        inst_loop: Loop,
+                        loop_analysis: &LoopAnalysis)
+                        -> DefPosition {
+    let mut closest_def_position = DefPosition::OutsideLoop();
+    for arg in func.dfg.inst_args(inst).into_iter() {
+        let new_def_position = match func.dfg.value_def(*arg) {
+            ValueDef::Arg(ebb_def, _) => {
+                match loop_analysis.base_loop_ebb(ebb_def) {
+                    None => DefPosition::OutsideLoop(),
+                    Some(loop_def) => {
+                        if inst_loop != loop_def {
+                            DefPosition::ParentLoop(loop_def)
+                        } else {
+                            DefPosition::InsideCurrentLoop()
+                        }
+                    }
                 }
-                // If there is a natural pre-header we insert new instructions just before the
-                // related jumping instruction (which is not necessarily at the end).
-                Some((_, last_inst)) => {
-                    pos = Cursor::new(&mut func.layout);
-                    pos.goto_inst(last_inst);
+            }
+            ValueDef::Res(inst_def, _) => {
+                match loop_analysis.base_loop_inst(inst_def, &func.layout) {
+                    None => DefPosition::OutsideLoop(),
+                    Some(loop_def) => {
+                        if inst_loop != loop_def {
+                            DefPosition::ParentLoop(loop_def)
+                        } else {
+                            DefPosition::InsideCurrentLoop()
+                        }
+                    }
+                }
+            }
+        };
+        // We now find some sort of minimum between `closest_def_position` and `new_def_position`.
+        match (closest_def_position.clone(), new_def_position) {
+            (DefPosition::InsideCurrentLoop(), _) |
+            (_, DefPosition::InsideCurrentLoop()) => {
+                closest_def_position = DefPosition::InsideCurrentLoop();
+                break;
+            }
+            (DefPosition::ParentLoop(_), DefPosition::OutsideLoop()) => break,
```

Comment:
Is a `break` out of the loop intended here? It seems like even if one of the args is outside the loop, later args could still be closer.